core/src/main/java/org/bitcoinj/core/Transaction.java

```diff
@@ -1413,67 +1412,64 @@ public synchronized Sha256Hash hashForWitnessSignature(
             int inputIndex,
             byte[] scriptCode,
             Coin prevValue,
-            byte sigHashType){
-        ByteArrayOutputStream bos = new ByteArrayOutputStream(255); // just a guess at an average tx length
-        try {
-            byte[] hashPrevouts = new byte[32];
-            byte[] hashSequence = new byte[32];
-            byte[] hashOutputs = new byte[32];
-            int basicSigHashType = sigHashType & 0x1f;
-            boolean anyoneCanPay = (sigHashType & SigHash.ANYONECANPAY.value) == SigHash.ANYONECANPAY.value;
-            boolean signAll = (basicSigHashType != SigHash.SINGLE.value) && (basicSigHashType != SigHash.NONE.value);
-
-            if (!anyoneCanPay) {
-                ByteArrayOutputStream bosHashPrevouts = new ByteArrayOutputStream(256);
-                for (TransactionInput input : this.inputs) {
-                    bosHashPrevouts.write(input.getOutpoint().hash().serialize());
-                    writeInt32LE(input.getOutpoint().index(), bosHashPrevouts);
-                }
-                hashPrevouts = Sha256Hash.hashTwice(bosHashPrevouts.toByteArray());
+            byte sigHashType) {
+        Sha256Hash hashPrevouts = Sha256Hash.ZERO_HASH;
+        Sha256Hash hashSequence = Sha256Hash.ZERO_HASH;
+        Sha256Hash hashOutputs = Sha256Hash.ZERO_HASH;
+        int basicSigHashType = sigHashType & 0x1f;
+        boolean anyoneCanPay = (sigHashType & SigHash.ANYONECANPAY.value) == SigHash.ANYONECANPAY.value;
+        boolean signAll = (basicSigHashType != SigHash.SINGLE.value) && (basicSigHashType != SigHash.NONE.value);
+
+        if (!anyoneCanPay) {
+            ByteBuffer bufHashPrevouts = ByteBuffer.allocate(this.inputs.size() * (Sha256Hash.LENGTH + 4));
```

Comment:
I think it would be helpful to implement the correct pre-allocation of buffer here for BAOS in a previous commit, since it is possible to do that and it will make reviewing easier if it is split into two commits.